File: packages/tslua-dcs-mission-types/scripts/export.bridge.lua
-- DCS Studio eval chunk. The driver substitutes the two quoted placeholders.
local namespace = [[NAMESPACE]]
local dcs_version = [[DCS_VERSION]]
local MAX_DEPTH, MAX_MEMBERS, MAX_SIZE = 12, 20000, 2000000
local members, size = 0, 0

local function die(message) error("tslua-dcs export " .. namespace .. ": " .. message) end
local function ts_quote(value)
  return '"' .. value:gsub('\\', '\\\\'):gsub('"', '\\"'):gsub('\n', '\\n'):gsub('\r', '\\r'):gsub('\t', '\\t'):gsub('[%z\1-\8\11\12\14-\31]', function(c) return string.format('\\u%04x', string.byte(c)) end) .. '"'
end
local function key_text(key)
  if type(key) == "string" then
    if key:match("^[A-Za-z_$][A-Za-z0-9_$]*$") then return key end
    return ts_quote(key)
  end
  if type(key) == "number" and key == math.floor(key) then return "[" .. key .. "]" end
  die("unsupported key type " .. type(key))
end
local function representable_key(key)
  return type(key) == "string" or (type(key) == "number" and key == math.floor(key))
end
local function sorted_keys(value)
  local keys, omitted = {}, 0
  for key in pairs(value) do
    if representable_key(key) then table.insert(keys, key) else omitted = omitted + 1 end
  end
  table.sort(keys, function(a,b) local ta,tb=type(a),type(b); if ta ~= tb then return ta < tb end; return tostring(a) < tostring(b) end)
  return keys, omitted
end
local function emit(lines, text)
  size = size + #text; if size > MAX_SIZE then die("generated source exceeds size limit") end
  table.insert(lines, text)
end
local function append(lines, value, depth, ancestors, path)
  if depth > MAX_DEPTH then die("depth limit at " .. path) end
  local indent = string.rep("\t", depth + 1)
  local keys, omitted = sorted_keys(value)
  if omitted > 0 then
    emit(lines, indent .. "/**")
    emit(lines, indent .. " * Omitted " .. omitted .. " non-representable Lua key(s) at " .. ts_quote(path) .. ";")
    emit(lines, indent .. " * their values are unknown because TypeScript cannot name Lua table keys.")
    emit(lines, indent .. " */")
  end
  for _, key in ipairs(keys) do
    if not (type(key) == "string" and key:sub(1,2) == "__") then
      members = members + 1; if members > MAX_MEMBERS then die("member limit at " .. path) end
      local item = rawget(value, key); local kind = type(item); local name = key_text(key)
      if kind == "function" then emit(lines, indent .. name .. "(...args: any[]): unknown;")
      elseif kind == "string" then emit(lines, indent .. name .. ": string;")
      elseif kind == "number" then emit(lines, indent .. name .. ": number;")
      elseif kind == "boolean" then emit(lines, indent .. name .. ": boolean;")
      elseif kind == "table" then
        if ancestors[item] then
          emit(lines, indent .. "/**")
          emit(lines, indent .. " * Cyclic reference at " .. ts_quote(path .. "." .. tostring(key)) .. ".")
          emit(lines, indent .. " */")
          emit(lines, indent .. name .. ": unknown;")
        else ancestors[item] = true; emit(lines, indent .. name .. ": {"); append(lines, item, depth + 1, ancestors, path .. "." .. tostring(key)); emit(lines, indent .. "};"); ancestors[item] = nil end
      else emit(lines, indent .. name .. ": unknown;") end
    end
  end
end

local root = rawget(_G, namespace)
if type(root) ~= "table" then die("root is not a table") end
if type(dcs_version) ~= "string" or dcs_version == "" then die("DCS version is missing") end
local comment_version = dcs_version:gsub("%*/", "* /"):gsub("[\r\n]", " ")
local lines = { "/**", " * @version " .. comment_version }
if not rawget(root, "className_") then table.insert(lines, " * @noSelf") end
table.insert(lines, " */"); table.insert(lines, "export interface _" .. namespace .. " {")
append(lines, root, 0, { [root] = true }, namespace); table.insert(lines, "}")
return { [namespace .. ".export.ts"] = table.concat(lines, "\n") .. "\n" }

File: packages/tslua-dcs-mission-types/tests/export-bridge.ts
/**
 * scripts/export.bridge.lua is the chunk `npm run export` evaluates inside DCS (through the project
 * mission bridge) to generate src/exports/*.export.ts from the live mission environment. These tests
 * run the same chunk on lua51 against synthetic namespaces, exactly as scripts/export.ts prepares it:
 * the two quoted placeholders are replaced by Lua string literals.
 */
import { describe, expect, test } from "@flying-dice/tslua-luatest";

function readBridge(): string {
	const [file, openError] = io.open("scripts/export.bridge.lua", "rb");
	if (!file) error(`cannot read the export bridge: ${openError}`);
	const source = file.read("*a") as string;
	file.close();
	return source;
}

const bridge = readBridge();

/** `%q` quoting is a valid Lua string literal for any bytes, like `lua()` in scripts/export.ts. */
function quote(value: string): string {
	return string.format("%q", value);
}

function exportNamespace(
	namespace: string,
	root: unknown,
	version = "2.9.29.27468",
): Record<string, string> {
	const [code] = string.gsub(
		string.gsub(bridge, "%[%[NAMESPACE%]%]", () => quote(namespace))[0],
		"%[%[DCS_VERSION%]%]",
		() => quote(version),
	);
	const [chunk, compileError] = loadstring(code, "=export.bridge.lua");
	if (!chunk) error(`bridge does not compile: ${compileError}`);
	const globals = _G as unknown as Record<string, unknown>;
	const previous = globals[namespace];
	globals[namespace] = root;
	try {
		return (chunk as () => Record<string, string>)();
	} finally {
		globals[namespace] = previous;
	}
}

function exported(namespace: string, root: unknown, version?: string): string {
	return exportNamespace(namespace, root, version)[`${namespace}.export.ts`];
}

describe("export bridge", () => {
	test("returns one file named after the namespace", () => {
		const files = exportNamespace("tsluaNamespace", { f: () => 1 });
		expect(Object.keys(files)).toEqual(["tsluaNamespace.export.ts"]);
	});

	test("a namespace becomes a @noSelf interface with typed members in sorted order", () => {
		const text = exported("tsluaNamespace", {
			zeta: () => 1,
			alpha: "text",
			count: 3,
			enabled: true,
			nested: { inner: () => 2, VALUE: 1 },
		});
		expect(text).toBe(
			[
				"/**",
				" * @version 2.9.29.27468",
				" * @noSelf",
				" */",
				"export interface _tsluaNamespace {",
				"\talpha: string;",
				"\tcount: number;",
				"\tenabled: boolean;",
				"\tnested: {",
				"\t\tVALUE: number;",
				"\t\tinner(...args: any[]): unknown;",
				"\t};",
				"\tzeta(...args: any[]): unknown;",
				"}",
				"",
			].join("\n"),
		);
	});

	test("a class table (with className_) is not @noSelf", () => {
		const text = exported("tsluaClass", {
			className_: "tsluaClass",
			getName: () => "x",
		});
		expect(text).not.toContain("@noSelf");
		expect(text).toContain(
			"\tclassName_: string;\n\tgetName(...args: any[]): unknown;\n",
		);
	});

	test("numeric keys come first as [n] and unusual string keys are quoted", () => {
		const root: Record<string | number, unknown> = {
			2: "two",
			1: "one",
			"with space": 1,
			'quo"te': 2,
			"back\\slash": 3,
			"tab\there": 4,
			"bell\u0007": 5,
			$valid_1: 6,
		};
		const text = exported("tsluaKeys", root);
		expect(text).toContain(
			[
				"\t[1]: string;",
				"\t[2]: string;",
				"\t$valid_1: number;",
				'\t"back\\\\slash": number;',
				'\t"bell\\u0007": number;',
				'\t"quo\\"te": number;',
				'\t"tab\\there": number;',
				'\t"with space": number;',
			].join("\n"),
		);
	});

	test("metamethod-style keys starting with __ are skipped", () => {
		const text = exported("tsluaHidden", {
			__index: {},
			__call: () => 1,
			visible: 1,
		});
		expect(text).not.toContain("__");
		expect(text).toContain("\tvisible: number;");
	});

	test("cycles are reported instead of followed", () => {
		const root: Record<string, unknown> = { name: "root" };
		const child: Record<string, unknown> = { back: root };
		root.child = child;
		child.self = child;
		const text = exported("tsluaCycle", root);
		expect(text).toContain(
			'\t\t * Cyclic reference at "tsluaCycle.child.back".',
		);
		expect(text).toContain("\t\tback: unknown;");
		expect(text).toContain(
			'\t\t * Cyclic reference at "tsluaCycle.child.self".',
		);
		expect(text).toContain("\t\tself: unknown;");
	});

	test("a table seen twice without a cycle is expanded both times", () => {
		const shared = { value: 1 };
		const text = exported("tsluaShared", { a: shared, b: shared });
		expect(text).toContain(
			"\ta: {\n\t\tvalue: number;\n\t};\n\tb: {\n\t\tvalue: number;\n\t};",
		);
	});

	test("keys TypeScript cannot name are counted in a comment", () => {
		const root = new LuaTable<AnyNotNil, unknown>();
		root.set(1.5, "fraction");
		root.set(true, "boolean");
		root.set({}, "table");
		root.set("kept", 1);
		const text = exported("tsluaOmitted", root);
		expect(text).toContain(
			'\t * Omitted 3 non-representable Lua key(s) at "tsluaOmitted";',
		);
		expect(text).toContain("\tkept: number;");
	});

	test("values of other Lua types become unknown", () => {
		const text = exported("tsluaOther", {
			routine: coroutine.create(() => undefined),
		});
		expect(text).toContain("\troutine: unknown;");
	});

	test("the version is sanitised so it cannot close the comment or add lines", () => {
		const text = exported("tsluaVersion", {}, "2.9 */ injected\r\nline");
		expect(text).toContain(" * @version 2.9 * / injected  line\n");
	});

	test("refuses a missing namespace", () => {
		expect(() => exported("tsluaMissing", undefined)).toThrow(
			"tslua-dcs export tsluaMissing: root is not a table",
		);
		expect(() => exported("tsluaString", "text")).toThrow(
			"root is not a table",
		);
	});

	test("refuses a missing DCS version", () => {
		expect(() => exported("tsluaNoVersion", {}, "")).toThrow(
			"tslua-dcs export tsluaNoVersion: DCS version is missing",
		);
	});

	test("stops at the depth limit", () => {
		const root: Record<string, unknown> = {};
		let current = root;
		for (let depth = 0; depth < 14; depth++) {
			const next: Record<string, unknown> = {};
			current.next = next;
			current = next;
		}
		expect(() => exported("tsluaDeep", root)).toThrow(
			"depth limit at tsluaDeep.next.next",
		);
	});

	test("stops at the member limit", () => {
		const root = new LuaTable<string, number>();
		for (let index = 1; index <= 20001; index++) root.set(`m${index}`, index);
		expect(() => exported("tsluaWide", root)).toThrow(
			"member limit at tsluaWide",
		);
	});

	test("stops at the generated size limit", () => {
		const root = new LuaTable<string, number>();
		const padding = string.rep("k", 200);
		for (let index = 1; index <= 10000; index++)
			root.set(`${padding}${index}`, index);
		expect(() => exported("tsluaLarge", root)).toThrow(
			"generated source exceeds size limit",
		);
	});
});
